src/pykeen/losses.py

```diff
@@ -1462,20 +1462,31 @@ def forward(
         num_entities: Optional[int] = None,
     ) -> torch.FloatTensor:  # noqa: D102
         # neg scores might be -inf for masked values -> get rid of those
-        mask = torch.isfinite(neg_scores)
-        neg_weights, neg_scores = neg_weights[mask], neg_scores[mask]
+        # mask = torch.isfinite(neg_scores)
+        # neg_weights, neg_scores = neg_weights[mask], neg_scores[mask]
         # create targets
         return self._reduction_method(
-            neg_weights
-            * functional.binary_cross_entropy_with_logits(
-                neg_scores,
-                apply_label_smoothing(torch.zeros_like(neg_scores), epsilon=label_smoothing, num_classes=num_entities),
-                reduction="none",
-            )
-        ) + functional.binary_cross_entropy_with_logits(
-            pos_scores,
-            apply_label_smoothing(torch.ones_like(pos_scores), epsilon=label_smoothing, num_classes=num_entities),
-            reduction=self.reduction,
+            torch.cat(
+                [
+                    neg_weights
+                    * functional.binary_cross_entropy_with_logits(
+                        neg_scores,
+                        apply_label_smoothing(
+                            torch.zeros_like(neg_scores), epsilon=label_smoothing, num_classes=num_entities
+                        ),
+                        reduction="none",
+                    ),
+                    functional.binary_cross_entropy_with_logits(
+                        pos_scores,
+                        apply_label_smoothing(
+                            torch.ones_like(pos_scores), epsilon=label_smoothing, num_classes=num_entities
+                        ),
+                        reduction="none",
+                    ),
+                ],
+                dim=-1,
+            ).sum(dim=-1)
+            / neg_weights.sum(dim=-1)
```

Comment:
Why would this make sense? 😅 the last dimension is the `num_negatives`-one; since they come from softmax, they should sum up to one along this dim 🤔

Reply:
Right, in the original implementation here:

https://github.com/DeepGraphLearning/torchdrug/blob/20f84170544d594a177e237ef5f3a1cadb2c61e6/torchdrug/tasks/reasoning.py#L89-L100

`new_weights` includes a scalar 1 for the positive score, so the sum over negatives (`new_weights[:, 1: ]`) does sum up to 1, but then we'd also add the original 1 from column 0, and overall the column sum is 2.

In our implementation, we have a `self.factor=0.5` here https://github.com/pykeen/pykeen/blob/62220dfe7e241115f7108bc87eb0749560da584d/src/pykeen/losses.py#L1312

but it weights the final reduced loss value, not rows in the batch.

Reply:
so maybe in our case the division should be ` torch.cat(...).sum(dim=-1 ) /  (1 + neg_weights.sum(dim=-1))`?

Reply:
it is strange that a division by two would make a difference as big as in https://github.com/pykeen/pykeen/issues/996 🤔

Reply:
I think the biggest difference is that in the old (current) version, `self._reduction_method()` for negatives is the same as `self.reduction` for positives - but in the OG formulation, we'd use `sum` for negatives (since weights come from the softmax), sum up scores in a row including a positive score, and then apply `mean` to the result. 
In the current form, if we specify `sum` as a reduction, then the positive scores would blow up

Reply:
Okay, so then maybe we should change that for all adversarial ones; For NSSA this should be just one line
```python
neg_loss = (neg_weights * neg_loss)
```
to
```python
neg_loss = (neg_weights * neg_loss).sum(dim=-1)
```
just before the reduction

Reply:
Or more generally, we always want a sum reduction along the softmax dim, before then using the "normal" reduction